### PR TITLE
Enable Flash Attention

### PR DESCRIPTION
**Pull request by jiqing-feng**
In CPU torch 2.5 we got flash attention for prefill, it's faster than varlen attention. We also apply SDPA when no past_key_values found.

It also use fused linear on gpt2 model.

### CODE REVIEW

**Comment by jiqing-feng** [comment]
Hi @sywangyi , please review these changes. Thanks!

File: optimum/exporters/ipex/modeling_utils.py [hunk 1]
```diff
@@ -585,9 +622,75 @@ def rope(self, *args, **kwargs):
...
+                is_causal=True,
+            )
+            self.use_sdpa = True
+        elif self.has_flash_attn(query):
+            # prefill, remove padding
+            attn_output = torch.empty_like(query)
+            seq_len_tensor = torch.cat((input_lens.new_tensor([0]), input_lens.cumsum(-1).int()))
+            PagedAttention.flash_attn_varlen_func(
+                attn_output,
+                query,
```
Comment (sywangyi) [inline optimum/exporters/ipex/modeling_utils.py:659]: xpu needs contiguous() for query as well

File: optimum/exporters/ipex/modeling_utils.py [hunk 2]
```diff
@@ -607,31 +710,18 @@ def forward(
...
-                True,
-                False,
-                None,
+            # prefill
+            attn_output = self.prefill_attn(
+                query, key_cache, value_cache, key, value, past_key_value, attention_mask, input_lens
             )
         else:
             # decode
+            attn_output = torch.empty_like(query)
```
Comment (sywangyi) [inline optimum/exporters/ipex/modeling_utils.py:724]: decode could use   PagedAttention.flash_attn_varlen_func as well

File: optimum/exporters/ipex/modeling_utils.py [hunk 3]
```diff
@@ -732,9 +836,11 @@ def rope(self, query, key, *args, **kwargs):
         return query, key
 
     def postprocess_attention_output(self, attn_output):
+        if self.use_sdpa:
+            attn_output = attn_output.transpose(1, 2).contiguous()
         attn_output = attn_output.reshape(-1, attn_output.shape[-2] * attn_output.shape[-1])
-        attn_output = self.c_proj(attn_output)
-        attn_output = self.resid_dropout(attn_output)
```
Comment (sywangyi) [inline optimum/exporters/ipex/modeling_utils.py:737]: why is drop removed here?

Comment (jiqing-feng) [reply optimum/exporters/ipex/modeling_utils.py:737]: Dropout will be disabled in inference mode, we can do linear add fusion without this dropout.

**Comment by jiqing-feng** [comment]
Hi @IlyasMoutawwakil. Do you mind reviewing this PR? Thanks!

**Review by IlyasMoutawwakil — APPROVED** [review]
LGTM

**Comment by jiqing-feng** [comment]
It has 20%+ speed-up for 1st token but the 2nd token latency increased, will fix it in the next PR.
